pkg/engine/loadtargets.go

```diff
@@ -82,99 +68,79 @@ func resolveSpec(i int, target kyvernov1.ResourceSpec, ctx engineapi.PolicyConte
 		}
-		if err != nil {
-			return nil, fmt.Errorf("failed to get target %s/%s %s/%s : %v", target.APIVersion, target.Kind, namespace, name, err)
-		}
-
-		return []resourceInfo{{unstructured: *obj, subresource: subresourceName, parentResourceGVR: parentResourceGVR}}, nil
-	}
-
-	if kubeutils.IsSubresource(apiResource.Name) {
-		apiVersion := metav1.GroupVersion{
-			Group:   parentAPIResource.Group,
-			Version: parentAPIResource.Version,
-		}.String()
-		objList, err := client.ListResource(context.TODO(), apiVersion, parentAPIResource.Kind, "", nil)
-		if err != nil {
-			return nil, err
-		}
-		var parentObjects []unstructured.Unstructured
-		for i := range objList.Items {
-			obj := objList.Items[i].DeepCopy()
-			if match(namespace, name, obj.GetNamespace(), obj.GetName()) {
-				parentObjects = append(parentObjects, *obj)
+		// we can use `GET` directly
+		if namespace != "" && name != "" && !wildcard.ContainsWildcard(namespace) && !wildcard.ContainsWildcard(name) {
+			var obj *unstructured.Unstructured
+			var err error
+			if namespace == "" {
+				obj, err = dyn.Get(context.TODO(), name, metav1.GetOptions{}, sub...)
+			} else {
+				obj, err = dyn.Namespace(namespace).Get(context.TODO(), name, metav1.GetOptions{}, sub...)
 			}
-		}
-
-		for i := range parentObjects {
-			parentObj := parentObjects[i]
-			subresourceName := strings.Split(apiResource.Name, "/")[1]
-			obj, err := client.GetResource(context.TODO(), parentObj.GetAPIVersion(), parentAPIResource.Kind, parentObj.GetNamespace(), parentObj.GetName(), subresourceName)
 			if err != nil {
 				return nil, err
 			}
-			parentResourceGVR := metav1.GroupVersionResource{
-				Group:    parentAPIResource.Group,
-				Version:  parentAPIResource.Version,
-				Resource: parentAPIResource.Name,
-			}
-			targetObjects = append(targetObjects, resourceInfo{unstructured: *obj, subresource: subresourceName, parentResourceGVR: parentResourceGVR})
-		}
-	} else {
-		// list all targets if wildcard is specified
-		objList, err := client.ListResource(context.TODO(), target.APIVersion, target.Kind, "", nil)
-		if err != nil {
-			return nil, err
-		}
-
-		for i := range objList.Items {
-			obj := objList.Items[i].DeepCopy()
-			if match(namespace, name, obj.GetNamespace(), obj.GetName()) {
-				targetObjects = append(targetObjects, resourceInfo{unstructured: *obj})
+			targetObjects = append(targetObjects, resourceInfo{unstructured: *obj, subresource: gvrs.SubResource, parentResourceGVR: metav1.GroupVersionResource(gvrs.GroupVersionResource)})
+		} else {
+			// we can use `LIST`
+			if gvrs.SubResource == "" {
+				list, err := dyn.List(context.TODO(), metav1.ListOptions{})
+				if err != nil {
+					return nil, err
+				}
+				for _, obj := range list.Items {
+					if match(namespace, name, obj.GetNamespace(), obj.GetName()) {
+						targetObjects = append(targetObjects, resourceInfo{unstructured: obj})
+					}
+				}
+			} else {
+				// we need to use `LIST` / `GET`
```

Comment:
What happens if a subresource cannot be List/Get?

Reply:
Of course if will fail (i didn't change that, it was already failing before)

Reply:
It makes sense to me the subresource must be list/get-able when it is a target. Got confused with this check https://github.com/kyverno/kyverno/pull/6544 - when the subresource acts as a trigger, it does not necessarily need to be Get-able.

Reply:
Definitely

Reply:
In addition we probably want to check it supports `update` too ;-)